### PR TITLE
[PR] tests: add KNN functions tests for `knn_L2sqr()` and `knn_inner_product()`

### PR DESCRIPTION
**Pull request by Coien-rr**
This commit adds comprehensive test functions for the `knn_L2sqr` and `knn_inner_product` algorithms. These tests verify the correctness of the distance computations and ensure that the KNN logic works as expected under various conditions.

- `test_knn_L2sqr`: Includes test cases to validate the squared Euclidean distance calculations.
- `test_KNN_inner_product`: Includes test cases to check the inner product similarity measure used in KNN.

### CODE REVIEW

File: tests/test_knn_functions.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                res_pairs.begin(),
+                res_pairs.end(),
+                [is_l2](const std::pair<float, int64_t>& a,
+                        const std::pair<float, int64_t>& b) {
+                    return is_l2 ? a.first < b.first : a.first > b.first;
+                });
+        for (size_t j = 0; j < k; ++j) {
+            ASSERT_EQ(res_pairs[j].second, indexes[q * k + j]);
+            // WARNING: There will be a very small fractional error :(
+            // ASSERT_FLOAT_EQ(res_pairs[j].first, distances[q * k + j]);
```
Comment (mnorris11) [inline tests/test_knn_functions.cpp:95]: Will this be a flaky test?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -36,14 +36,14 @@ set(FAISS_TEST_SRC
...
+  test_knn_functions.cpp
 )
 
 add_executable(faiss_test ${FAISS_TEST_SRC})
 
 include(../cmake/link_to_faiss_lib.cmake)
 
 link_to_faiss_lib(faiss_test)
-
 target_link_libraries(faiss_test PUBLIC faiss_example_external_module)
```
Comment (mnorris11) [inline tests/CMakeLists.txt:47]: On latest code, this line is deleted

Comment (Coien-rr) [reply tests/CMakeLists.txt:47]: this conflict has resolved

File: tests/test_knn_functions.cpp (hunk 1 above)
Comment (Coien-rr) [reply tests/test_knn_functions.cpp:95]: I rewrote the test, but the results appear to be in error

```cpp
            ASSERT_EQ(res_pairs[j].second, indexes[q * k + j]);
            // WARNING: There will be a very small fractional error :(
            // ASSERT_FLOAT_EQ(res_pairs[j].first, distances[q * k + j]);
            if (is_l2) {
                ASSERT_FLOAT_EQ(
                        faiss::fvec_L2sqr(
                                queries + q * dim,
                                db + res_pairs[j].second * dim,
                                dim),
                        distances[q * k + j]);
            }
```

but i got error follow 
```
Test project /home/byte/Code/ANNS/faiss/build
    Start 115: TestKNNFunctions.knn_L2sqr
1/1 Test #115: TestKNNFunctions.knn_L2sqr .......***Failed    0.02 sec
WARNING clustering 1000 points to 40 centroids: please provide at least 1560 training points
Running main() from /home/byte/Code/ANNS/faiss/build/_deps/googletest-src/googletest/src/gtest_main.cc
Note: Google Test filter = TestKNNFunctions.knn_L2sqr
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from TestKNNFunctions
[ RUN      ] TestKNNFunctions.knn_L2sqr
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
/home/byte/Code/ANNS/faiss/tests/test_knn_functions.cpp:105: Failure
Expected equality of these values:
  faiss::fvec_L2sqr( queries + q * dim, db + res_pairs[j].second * dim, dim)
    Which is: 0
  distances[q * k + j]
    Which is: 2.2888184e-05
[  FAILED  ] TestKNNFunctions.knn_L2sqr (4 ms)
[----------] 1 test from TestKNNFunctions (4 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (4 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestKNNFunctions.knn_L2sqr

 1 FAILED TEST


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.02 sec

The following tests FAILED:
        115 - TestKNNFunctions.knn_L2sqr (Failed)
Errors while running CTest
```
that means even though the two points have the correct nearest neighbor relationship, the knn function's distance calculation is subject to error

1. **Is it possible to get an error with the knn function in the library?**
2. **Or there might be a problem with the generated random data in my test?**

File: tests/CMakeLists.txt (hunk 2 above)
Comment (mnorris11) [reply tests/CMakeLists.txt:47]: Sorry if I was not clear: this line should be deleted. The latest code on master deletes it, so we should not re-add it.
(I am still reading through the test, sorry if there is some delay there)

Comment (Coien-rr) [reply tests/CMakeLists.txt:47]: oh my god, i maybe midunderstood your meaning, I thought it was a blank line problem.
Sadddddddddd :(

**Review by mnorris11 — COMMENTED** [review]
Thanks for adding the tests! I added some comments, mostly just about matching the repo's style, and one about keeping the float assertion, just accounting for the error.

**Comment by Coien-rr** [comment]
I received your reviews suggestion and will fix the code style issue as soon as possible!
thanks bro @mnorris11

File: tests/test_knn_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                    top_k,
+                    distances.get(),
+                    indexes.get(),
+                    true);
+        }
+    }
+}
+
+// TODO:
+TEST(TestKNNFunctions, knn_inner_product) {
```
[the remark this answers is not in the document]
Comment (Coien-rr) [reply tests/test_knn_functions.cpp:134]: yes, it is complete test, i forgot remove it after dev, sorry

File: tests/test_knn_functions.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+#include <cstdint>
+#include <memory>
+#include <random>
+
+#include "faiss/utils/distances.h"
+
+static const unsigned NPTS{1000};
+static const unsigned DIM{128};
+static const std::array<unsigned, 3> TOP_K_LIST{1, 50, 100};
+static const std::array<unsigned, 2> NUMS_QUREY_LIST{10, 30};
```
[the remark this answers is not in the document]
Comment (Coien-rr) [reply tests/test_knn_functions.cpp:21]: ok, i got this code style issue

**Comment by mdouze** [comment]
I am not sure that this test is useful because the knn functions are tested through the IndexFlat tests here: 

https://github.com/facebookresearch/faiss/blob/main/tests/test_index.py#L27

These tests are probably better because: 
- the reference implementation is in numpy instead of being another C++ implementation that could have bugs of its own
- they are in python so more compact and easier to change 
- they cover range search as well. 

As a side note, we will probably need tests for different d's because soon fused kernels specialized for some d's will be added to Faiss.

**Comment by Coien-rr** [comment]
i tried to add various dims && npts for knn test. but i got that question:

**when it has two neighbors with same distancce, `Faiss` doesn't seem to be able to prioritize.**

image

I think that at the same distance, the node with the smaller id should probably be in front of the point with the larger id ( if you guys agree with that opinion

But `faiss` doesn't seem to have specifically dealt with the case of equal distances.

1. topk == 1 case:
https://github.com/facebookresearch/faiss/blob/1a799d014b07aea1fca6788691078054f7865f4c/faiss/impl/ResultHandler.h#L126-L133

2. topk > 1  case: 
https://github.com/facebookresearch/faiss/blob/1a799d014b07aea1fca6788691078054f7865f4c/faiss/impl/ResultHandler.h#L262-L278

**Comment by mnorris11** [comment]
Apologies @Coien-rr , I didn't catch that the IndexFlat tests were already testing this. We'll go ahead and close it for now.

To answer about the ids: ids can be anything, completely unrelated to how similar the vectors are. So preferring a smaller or larger one is not needed.

**Pull request closed without merging**